flight/PiOS/STM32/inc/pios_dio.h

```diff
@@ -0,0 +1,313 @@
+/**
+ ******************************************************************************
+ * @file       pios_dio.h
+ * @author     dRonin, http://dRonin.org/, Copyright (C) 2017
+ * @addtogroup PIOS PIOS Core hardware abstraction layer
+ * @{
+ * @addtogroup PIOS_DIO Digital IO subsystem
+ * @{
+ *
+ * @brief Provides an interface to configure and efficiently use GPIOs.
+ *
+ * @details DIO holds a GPIO specifier (bank, pin) in a pointer-width integer.
+ * Inline functions in this module provide an efficient interface by which to
+ * set/reset/toggle/read the GPIO.
+ *
+ * Partially inspired by the iotag subsystem in CleanFlight/BetaFlight.
+ *
+ * To use this module, first store a GPIO into a dio_tag_t -- e.g.
+ *
+ * dio_tag_t t = GPIOA_DIO(4);
+ *
+ * for pin A4 on STM32.   Then use the below functions to access it.
+ *****************************************************************************/
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, see <http://www.gnu.org/licenses/>
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#ifndef _PIOS_DIO_H
+#define _PIOS_DIO_H
+
+#ifdef STM32F10X_MD
+#error no STM32F1 support yet
+#endif
+
+#ifdef STM32F4xx
+enum dio_drive_strength {
+	DIO_DRIVE_WEAK = 0,
+	DIO_DRIVE_LIGHT,
+	DIO_DRIVE_MEDIUM,
+	DIO_DRIVE_STRONG
+};
+#else
+enum dio_drive_strength {
+	DIO_DRIVE_WEAK = 0,
+	DIO_DRIVE_LIGHT = 0,
+	DIO_DRIVE_MEDIUM = 1,
+	DIO_DRIVE_STRONG = 3
+};
+#endif
+
+enum dio_pin_function {
+	DIO_PIN_INPUT = 0,
+	DIO_PIN_OUTPUT = 1,
+	DIO_PIN_ALTFUNC = 2,
+	DIO_PIN_ANALOG = 3
+};
+
+enum dio_pull {
+	DIO_PULL_NONE = 0,
+	DIO_PULL_UP,
+	DIO_PULL_DOWN
+};
+
+typedef uintptr_t dio_tag_t;
+
+/**
+ * @brief Configures a GPIO as alternate function.
+ * Note that to configure drive strength, pullups, etc, DIO_output or
+ * DIO_input should be called first.
+ *
+ * @param[in] t The GPIO specifier tag (created with e.g. GPIOA_DIO(3) )
+ * @param[in] alt_func The alternate function specifier.
+ */
+static inline void DIO_altfunc(dio_tag_t t, int alt_func);
```

Comment:
need to decide what we're doing with prefixen in general.  PIOS etc are uppercase prefixes still.. I don't know.